lib/inc/internal/sources/dmi_source.hpp

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <internal/vm.hpp>
+#include <memory>
+#include <string>
+
+namespace whereami { namespace sources {
+
+    /**
+     * Stores information collected via DMI
+     */
+    struct dmi_data
+    {
+        std::string bios_vendor;
+        std::string board_manufacturer;
+        std::string board_product_name;
+        std::string manufacturer;
+        std::string product_name;
+    };
+
+    /**
+     * Base DMI data source
+     */
+    class dmi_base
+    {
+    public:
+        dmi_base() {}
+        virtual ~dmi_base() {}
+        /**
+         * Retrieve the BIOS vendor
+         * @return The BIOS vendor's name
+         */
+        std::string bios_vendor() const;
+        /**
+         * Retrieve the board manufacturer
+         * @return The board manufacturer's name
+         */
+        std::string board_manufacturer() const;
+        /**
+         * Retrieve the board product name
+         * @return The board product name
+         */
+        std::string board_product_name() const;
+        /**
+         * Retrieve the system manufacturer
+         * @return The system manufacturer
+         */
+        std::string manufacturer() const;
+        /**
+         * Retrieve the product name
+         * @return The product name
+         */
+        std::string product_name() const;
+    protected:
+        /**
+         * Collected data for this machine based on DMI information
+         */
+        std::unique_ptr<dmi_data> data_;
+        /**
+         * Attempt to collect data about this machine using DMI
+         */
+        virtual void collect_data() = 0;
```

Comment:
I don't think this needs to be part of the base interface.

Since you can't call virtual functions from constructors, each subclass will have to call its own collect_data function in their own constructor (which you do correctly in the code below). That means nothing should be calling this method via a `dmi_base` reference.